lib/Target/DebugInfo/EmitHGLDD.cpp

```diff
@@ -0,0 +1,369 @@
+  findAndEmitLoc(json, "hgl_loc", variable->loc, false);
+  findAndEmitLoc(json, "hdl_loc", variable->loc, true);
+
+  if (auto value = variable->value) {
+    StringAttr portName;
+    auto *defOp = value.getParentBlock()->getParentOp();
+    auto module = dyn_cast<hw::HWModuleOp>(defOp);
+    if (!module)
+      module = defOp->getParentOfType<hw::HWModuleOp>();
+    if (module) {
+      if (auto arg = dyn_cast<BlockArgument>(value)) {
+        portName = dyn_cast_or_null<StringAttr>(
+            module.getInputNames()[arg.getArgNumber()]);
+      } else if (auto wireOp = value.getDefiningOp<hw::WireOp>()) {
+        portName = wireOp.getNameAttr();
+      } else {
+        for (auto &use : value.getUses()) {
+          if (auto outputOp = dyn_cast<hw::OutputOp>(use.getOwner())) {
+            portName = dyn_cast_or_null<StringAttr>(
+                module.getOutputNames()[use.getOperandNumber()]);
+            break;
+          }
+        }
+      }
+    }
+    if (auto intType = dyn_cast<IntegerType>(value.getType())) {
+      json.attribute("type_name", "logic");
+      if (intType.getIntOrFloatBitWidth() != 1) {
+        json.attributeArray("packed_range", [&] {
+          json.value(intType.getIntOrFloatBitWidth() - 1);
+          json.value(0);
+        });
+      }
+    }
+    if (portName) {
+      json.attributeObject(
+          "value", [&] { json.attribute("sig_name", portName.getValue()); });
+    }
+  }
+
+  json.objectEnd();
+}
+
+//===----------------------------------------------------------------------===//
+// Output Splitting
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+/// Contextual information for HGLDD emission shared across multiple HGLDD
+/// files. This struct is used to determine an initial split of debug info files
+/// and to distribute work.
+struct Emitter {
+  DebugInfo di;
+  SmallVector<FileEmitter, 0> files;
+
+  Emitter(Operation *module, const EmitHGLDDOptions &options);
+};
+
+} // namespace
+
+Emitter::Emitter(Operation *module, const EmitHGLDDOptions &options)
+    : di(module) {
+  // Group the DI modules according to their emitted file path. Modules that
+  // don't have an emitted file path annotated are collected in a separate
+  // group.
+  MapVector<StringAttr, FileEmitter> groups;
+  for (auto [moduleName, module] : di.moduleNodes) {
+    StringAttr hdlFile;
+    if (module->op)
+      if (auto fileLoc = findBestLocation(module->op->getLoc(), true))
+        hdlFile = fileLoc.getFilename();
+    groups[hdlFile].modules.push_back(module);
+  }
+
+  // Determine the output file names and move the emitters into the `files`
+  // member.
+  files.reserve(groups.size());
+  for (auto &[hdlFile, emitter] : groups) {
+    emitter.options = &options;
+    emitter.hdlFile = hdlFile;
+    emitter.outputFileName = options.outputDirectory;
+    StringRef fileName = hdlFile ? hdlFile.getValue() : "global";
```

Comment:
I see, null `StringAttr` is handled here.

Reply:
Yeah that's definitely worth a comment!